rust/datafusion/src/execution/physical_plan/memory.rs

```diff
@@ -0,0 +1,162 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Execution plan for reading in-memory batches of data
+
+use std::sync::{Arc, Mutex};
+
+use crate::error::Result;
+use crate::execution::physical_plan::{BatchIterator, ExecutionPlan, Partition};
+use arrow::datatypes::Schema;
+use arrow::record_batch::RecordBatch;
+
+/// Execution plan for reading in-memory batches of data
+pub struct MemoryExec {
+    /// The partitions to query
+    partitions: Vec<Vec<RecordBatch>>,
+    /// Schema representing the data after the optional projection is applied
+    schema: Arc<Schema>,
+    /// Optional projection
+    projection: Option<Vec<usize>>,
+}
+
+impl ExecutionPlan for MemoryExec {
+    /// Get the schema for this execution plan
+    fn schema(&self) -> Arc<Schema> {
+        self.schema.clone()
+    }
+
+    /// Get the partitions for this execution plan. Each partition can be executed in parallel.
+    fn partitions(&self) -> Result<Vec<Arc<dyn Partition>>> {
+        let partitions = self
+            .partitions
+            .iter()
+            .map(|vec| {
+                Arc::new(MemoryPartition::new(
+                    vec.clone(),
+                    self.schema.clone(),
+                    self.projection.clone(),
+                )) as Arc<dyn Partition>
+            })
+            .collect();
+        Ok(partitions)
+    }
+}
+
+impl MemoryExec {
+    /// Create a new execution plan for reading in-memory record batches
+    pub fn try_new(
+        partitions: &Vec<Vec<RecordBatch>>,
+        schema: Arc<Schema>,
+        projection: Option<Vec<usize>>,
+    ) -> Result<Self> {
+        Ok(Self {
+            partitions: partitions.clone(),
+            schema,
+            projection,
+        })
+    }
+}
+
+/// Memory partition
+struct MemoryPartition {
+    /// Vector of record batches
+    data: Vec<RecordBatch>,
+    /// Schema representing the data
+    schema: Arc<Schema>,
+    /// Optional projection
+    projection: Option<Vec<usize>>,
+}
+
+impl MemoryPartition {
+    /// Create a new in-memory partition
+    fn new(
+        data: Vec<RecordBatch>,
+        schema: Arc<Schema>,
+        projection: Option<Vec<usize>>,
+    ) -> Self {
+        Self {
+            data: data.clone(),
```

Comment:
would it be better to take ownership of the vec here without incurring the clone?

Reply:
Good point. Thanks. Updated.